install gtp2ogs and node_modules locally (no global -g), default to devel, add package-lock.json

wonderingabout opened the pull request:
edit: rebased with no change since the last few days
rewrote commit notes cleanly

sanderland commented on README.md:
```diff
@@ -2,179 +2,128 @@
 
-For linux (preferably as sudo):
+Before you can play on official [OGS server](https://online-go.com/) server, OGS admins
+ will often ask that you run tests to make sure everything is running fine.
```
I did not find this to be true

wonderingabout replied:
indeed they dont always ask it, but they often do, especially if they see the new bot admin is fairly new
so its indeed not always true but often

sanderland commented on README.md:
```diff
@@ -2,179 +2,128 @@
 
-- for linux: [3A3) Recommended: Upgrade gtp2ogs.js from old branch to “devel” branch (latest)](https://github.com/wonderingabout/gtp2ogs-tutorial/blob/master/docs/3A3-linux-optional-upgrade-to-devel.md)
-- for windows: [3B3) Recommended: Upgrade gtp2ogs from old branch to devel (latest) branch](https://github.com/wonderingabout/gtp2ogs-tutorial/blob/master/docs/3B3-windows-optional-upgrade-to-devel.md)
+After that, it is mandatory to also yout AI settings. To do that, add a `--`
```
this sentence is a word

sanderland replied:
and I would rewrite in terms of 'the command and arguments used to run your bot' rather than talking about settings (which could be anywhere)

wonderingabout replied:
indeed, the wording is wrong also :+1:

sanderland commented on README.md:
````diff
@@ -2,179 +2,128 @@
+When all testing is OK, you can remove the `--beta` and play in the real OGS.
 
-```Shell
-pushd C:\Program Files\nodejs && node.exe C:\Users\yourwindowsusername\AppData\Roaming\npm\node_modules\gtp2ogs\gtp2ogs.js --username <yourbotusername> --apikey <apikey> <gtp2ogsargument1> <gtp2ogsargument2> -- C:\Users\path\to\your\ai\executable.exe <botargument1> <botargument2>
````
where does pushd come from?

wonderingabout replied:
i dont understand this question

sanderland commented on README.md:
```diff
@@ -2,179 +2,128 @@
-and [apikey](https://github.com/online-go/gtp2ogs/blob/devel/docs/OPTIONS-LIST.md#apikey)
-first.
+`
+pushd C:\Program Files\nodejs\node.exe C:\Users\yourusername\gtp2ogs-node\node_modules\gtp2ogs\gtp2ogs.js --username <yourbotusername> --apikey <apikey> <gtp2ogsargument1> <gtp2ogsargument2> -- C:\Users\path\to\your\ai\executable.exe <botargument1> <botargument2>
```
i would make these example command as small as possible

wonderingabout replied:
i removed a few commands (bans (since we dont always use these), maxconnectedgames and maxconnectedgamesperuser (since we provide defaults for these and are unlikely to want to change the defaults), i also added ogspv to highlight that it exists

for the other options i thought it would be a good way to illustrate a real use case of gtp2ogs to show the range of possibilities to a new user

sanderland commented on README.md:
```diff
@@ -2,179 +2,128 @@
+ [Submit-Move Button](https://github.com/wonderingabout/gtp2ogs-tutorial#important-submit-move-button).
 
-### Show winrate and variations ingame
+## Show winrate and variations ingame
```
why is this a section here at all, rather than somewhere in options?

wonderingabout replied:
because it's an important add to gtp2ogs, so i think it's worth giving it extra visibility (just a link to the option details)

sanderland commented on README.md:
```diff
@@ -2,179 +2,128 @@
+
+## Developer Notes
+
+You can find notes for developers [in DEV.md](/docs/DEV.md).
```
[here](/docs/DEV.md)

wonderingabout replied:
i dont understand, the link seems to correctly redirect to `/docs/DEV.md` on my local branch

sanderland replied:
I mean, link with the word 'here'

wonderingabout replied:
ah i see, fixed :+1:

sanderland commented on docs/INSTALL-GUIDE-LINUX.md:
````diff
@@ -0,0 +1,80 @@
+For example:
+
+```Shell
+curl -o- https://raw.githubusercontent.com/nvm-sh/nvm/v0.35.3/install.sh | bash
````
can this not be done with apt-get or something?

wonderingabout replied:
no, apt only provides nodejs 10.19.xx (it was the old method by the way) default, and if we add the curl from nodejs repo we get package conflicts and broken dependencies

wonderingabout replied:
i added a note for it to be clear

sanderland commented on docs/INSTALL-GUIDE-LINUX.md:
```diff
@@ -0,0 +1,80 @@
+You can go to `~/gtp2ogs-node/node_modules/gtp2ogs`, delete all existing
+ files (make a copy backup just in case), then either:
+
+- download ZIP from github [the devel branch](https://github.com/online-go/gtp2ogs/tree/devel),
```
why not git clone?

wonderingabout replied:
it is possible to use git (as explained below), but for beginners they may find it easier to copy paste the files directly

wonderingabout replied:
but indeed i forgot to add the git clone command line, added :+1:

sanderland commented on docs/INSTALL-GUIDE-WINDOWS.md:
```diff
@@ -0,0 +1,55 @@
+- download ZIP from github [the devel branch](https://github.com/online-go/gtp2ogs/tree/devel),
+ then go inside `C:\Users\yourusername\gtp2ogs-node\node_modules\gtp2ogs` folder,
+ and paste all the new devel branch files.
+- or another use `git` if you're familiar with it to install devel from online-go/gtp2ogs or
```
typos

wonderingabout replied:
fixed, thanks :+1:

wonderingabout commented:
> some suggestions

> some suggestions

big thanks, they are for the most really accurate and relevant, thanks! :100: 

feel free to add other feedback

Dorus commented on README.md:
````diff
@@ -2,179 +2,112 @@
-node /usr/lib/node_modules/gtp2ogs/gtp2ogs.js --username <yourbotusername> --apikey <apikey> <gtp2ogsargument1> <gtp2ogsargument2> -- /replace/with/full/path/to/your/ai/runfile.file <botargument1> <botargument2>
-```
+Then, choose the wanted `options` from the above list of all options section
+ (no need to use all these options, only the ones you want), and separate each option with a space.
````
It might be better to explain options can be set with command line arguments, and those have to be separated by spaces. This explanation could also come earlier in the introduction i think.

I'm not even sure you have to mention argument have to be space separated, this is pretty standard. Just mention arguments can be used to set the various options.

wonderingabout replied:
exactly what i did!

then i'll also remove it, it is obvious from the examples anyways so better not overload the readme

Dorus commented on README.md:
````diff
@@ -2,179 +2,112 @@
-```Shell
-pushd C:\Program Files\nodejs && node.exe C:\Users\yourwindowsusername\AppData\Roaming\npm\node_modules\gtp2ogs\gtp2ogs.js --username <yourbotusername> --apikey <apikey> <gtp2ogsargument1> <gtp2ogsargument2> -- C:\Users\path\to\your\ai\executable.exe <botargument1> <botargument2>
-```
+After that, specify your bot command (and its arguments) using a , add a ` -- `
````
> After that, specify your bot command (and its arguments) using a , add a ` -- `

I'm not sure what you mean here, there is no need for a comma in there, i think the words ` using a ` have to be removed here.

wonderingabout replied:
yes it was very messy, fixed

Dorus commented on README.md:
````diff
@@ -2,179 +2,112 @@
-```
+After that, specify your bot command (and its arguments) using a , add a ` -- `
+ (with one space before and one space after the `--` ), to separate `options`
+ from your bot path and `bot arguments`.
````
Why is bot arguments in code style?

Did you mean to articulate it? I might prefer italic `from your *bot path* and *bot arguments*.`
>  from your *bot path* and *bot arguments*.

wonderingabout replied:
actually i wrote it in normal text

Dorus commented on README.md:
```diff
@@ -2,179 +2,112 @@
-  
-note 3: to play on [beta OGS server](https://beta.online-go.com/) instead of the
-[OGS server](https://online-go.com/), add the `--beta` argument
+`path/to/node_executable path/to/gtp2ogs.js --username <yourbotusername> --apikey <apikey> --option1 <argument1> --option2 <argument2> -- path/to/your/ai/executable <botoption1> <botargument1> <botoption2> <botargument2>`
```
You should probably make this a code block, it's too long for inline style.

\`\`\`Shell
[code]
\`\`\`

Dorus replied:
This also count for the other blocks below.

wonderingabout replied:
it was a code block at first, but then i thought it was not clearly visible in one glance

isnt it better to be able to see it all without scrolling?

Dorus commented on README.md:
```diff
@@ -2,179 +2,112 @@
-to separate `<gtp2ogsarguments>` from your bot path and `<botarguments>`, as
-shown in
-[Most common usage earlier](#4-most-common-usage--start-gtp2ogsjs-using-nodejs)
+`C:\Program Files\nodejs\node.exe C:\Users\yourusername\gtp2ogs-node\node_modules\gtp2ogs\gtp2ogs.js --apikey 5rg46ze84f68e4g6v4e4g68es4g8 --username somebot --debug --ogspv PhoenixGo --noclock --unrankedonly --minrank 1d --noautohandicapranked --maxhandicap 0 --boardsizes 19 --komis 7.5 --speeds live --maxmaintimelive 600 --maxperiodsliveranked 5 --maxperiodsliveunranked 10 --maxperiodtimelivera ...
```
No need to use `C:\Program Files\nodejs\` if you installed node.js and have it in your path.

wonderingabout replied:
then just node.exe ?

Dorus commented on README.md:
```diff
@@ -2,179 +2,112 @@
-simultaneous games, corrqueue, etc.)
-- choose your bot-specific options (for example `-v 1600` for leela-zero,
-etc.)
+ (example: rank, boardsize, blitz/live/correspondence, etc.), or to
```
board size

Dorus commented on README.md:
```diff
@@ -2,179 +2,112 @@
+ [Submit-Move Button](https://github.com/wonderingabout/gtp2ogs-tutorial#important-submit-move-button).
 
-**gtp2ogs has default support for showing ingame winrate and variations for some AI.**
+## Show winrate and variations ingame
```
in game
or
during games

wonderingabout replied:
in games?

Dorus commented on README.md:
```diff
@@ -2,179 +2,112 @@
+gtp2ogs has default support for showing ingame winrate and variations for some AI,
+ see: [--ogspv](/docs/OPTIONS-LIST.md/#ogspv)
 
 ## Community Involvment
```
Involvement

wonderingabout replied:
unrelated but it's actually not a default (not auto enabled) but a "native" feature

fixed typo too

Dorus commented on README.md:
```diff
@@ -2,179 +2,112 @@
-- You may be interested in joining the Discord chat:
-In leela zero discord (`#bots` channel): <https://discord.gg/HZ23Cp9>
+ to report issues.
+- Contributing in most welcome, if you want to add features or submit fixes
```
Shouldn't that be `is most welcome`?

wonderingabout replied:
yes, fixed typo

Dorus commented on docs/INSTALL-GUIDE-LINUX.md:
```diff
@@ -0,0 +1,84 @@
+
+gtp2ogs will be installed in `~/gtp2ogs-node/node_modules/gtp2ogs`
+
+## 3. Recommended: Upgrade to devel branch
```
Why first download old gtp2ogs if you want to switch to dev anyway? I would skip step 2 if you are planing to do step 3 anyway.
That would also eliminate the need to delete the `node_modules/gtp2ogs` folder at this step.

wonderingabout replied:
yes!!!!

but how do you install latest devel with npm ?
(for linux and windows)

Dorus commented on docs/INSTALL-GUIDE-LINUX.md:
```diff
@@ -0,0 +1,84 @@
+
+## 3. Recommended: Upgrade to devel branch
+
+This step is **highly recommended**.
```
Note we're recommending people to switch to a beta version here.

Not per sé bad, it has a lot of new good features, but we should probably put a note about that here.

Also makes me think if we should add some branching strategy where users can use `master` and we work in `develop` branch or so. Only merging to master once we know it's stable for a bit.

Dorus replied:
ps. this comment also count for the windows guide.

wonderingabout replied:
yes, i'll explain that it is beta

can you explain the branching strategy in detail?

wonderingabout replied:
is it how to do it ?

wonderingabout replied:
is it how to do it?
https://stackoverflow.com/a/39732501

for example?
```Shell
npm install git://github.com/online-go/gtp2ogs.git#devel --save
```

wonderingabout replied:
oh wow it works!!

![npm install devel](https://user-images.githubusercontent.com/38690718/81476116-48660900-9210-11ea-91a4-8f478be36bb8.png)

its not a git repo so we cant clone it, but at least we have latest devel natively

wonderingabout replied:
found an even better way (now we clone from github and can use git in the repo), but we need to add node_modules and package-lock.json to git ignore

```Shell
git clone -b devel https://github.com/online-go/gtp2ogs && cd gtp2ogs && npm i
```

![npm install clone](https://user-images.githubusercontent.com/38690718/81476604-0ab6af80-9213-11ea-997e-cd1c66f95565.png)

from: https://stackoverflow.com/a/45847287

wonderingabout commented on README.md:
```diff
@@ -2,179 +2,112 @@
-This javascript tool allows all bots/AI that support
-[GTP (Go Text Protocol)](https://senseis.xmp.net/?GoTextProtocol)
-to communicate with [OGS (Online-Go.com Server)](https://online-go.com/).
+ This javascript tool allows all bots/AI that support
```
the spaces on lines 5,9,12 are indeed typos, the other ones at the beginning of a newline are part of markdownlint style

wonderingabout commented on README.md:
```diff
@@ -2,179 +2,112 @@
 [for windows and linux, with screenshots, examples, and detailed explanations](https://github.com/wonderingabout/gtp2ogs-tutorial)
 
-## Quick Guide
+## Quick Install Guide
```
indeed, it seems a more accurate word! https://wiki.archlinux.org/index.php/installation_guide

wonderingabout commented on README.md:
```diff
@@ -2,179 +2,112 @@
+## Quick Install Guide
 
-### 1. install node.js and npm
+Installing node.js will also install:
```
yes, how about this `The "node.js" package is bundled with:` ?

wonderingabout commented on README.md:
```diff
@@ -2,179 +2,112 @@
-- for windows:
-
-Open a node.js command prompt as admin, then:
+## Options
```
fixed

wonderingabout commented on README.md:
```diff
@@ -2,179 +2,112 @@
+### Options settings
 
-### 4. Most common usage: start gtp2ogs.js using nodejs
+Before putting `options`, you have to first put the mandatory options:
```
yes, i thought the same but forgot, thanks

wonderingabout commented on README.md:
````diff
@@ -2,179 +2,112 @@
-```Shell
-node /usr/lib/node_modules/gtp2ogs/gtp2ogs.js --username <yourbotusername> --apikey <apikey> <gtp2ogsargument1> <gtp2ogsargument2> -- /replace/with/full/path/to/your/ai/runfile.file <botargument1> <botargument2>
-```
+Then, choose the wanted `options` from the above list of all options section
````
actually i simpified it now and moved the options-list link where we actually need it

chronological and logical :100:

Dorus commented on .gitignore:
```diff
@@ -36,5 +36,6 @@ nosetests.xml
 .pydevproject
 
 node_modules
+package-lock.json
```
> This file is intended to be committed into source repositories

https://docs.npmjs.com/configuring-npm/package-lock-json.html

wonderingabout replied:
i see, then i should remove it again i guess, thanks

wonderingabout replied:
fixed

Dorus replied:
Yes, and once you've done a clean `npm install`, you should also commit this file.

wonderingabout replied:
should we add this information in the readme?

roy7 commented:
Any new or outstanding issues before merge? @Dorus @sanderland

wonderingabout commented:
should we commit package-lock.json as well in this pr as dorus suggested in https://github.com/online-go/gtp2ogs/pull/260#discussion_r422624737?

Dorus commented on docs/INSTALLATION-LINUX.md:
```diff
@@ -0,0 +1,59 @@
+
+gtp2ogs will be installed in `~/gtp2ogs/`.
+
+node_modules folder and package-lock.json are included in .gitignore
```
package lock is not in gitigore right?

I'm not even sure on the relevance of this info here. It's pretty standard to have node_modules in gitignore.

wonderingabout replied:
yes you're right, it was my misunderstanding

I removed the gitignored of package-lock.json
and committed package-lock.json in this pr

Dorus commented:
@roy7 This seems ok to me, can be merged when meta updates the final text i just mentioned.

wonderingabout commented:
> @roy7 This seems ok to me, can be merged when meta updates the final text i just mentioned.

thanks, i did and added a note about license for developers, is it fine?
https://github.com/wonderingabout/gtp2ogs/blob/update-doc-node-local-install-nvm/docs/DEV.md#about-license

as we discussed yesterday on discord, some big projects have different licenses for different files, so i thought this precision is good to add

is it ok?

Dorus commented:
@wonderingabout I'm not sure it's accurate that all files are under the same license. The code files at least are and should get a header for that.

wonderingabout commented:
> @wonderingabout I'm not sure it's accurate that all files are under the same license. The code files at least are and should get a header for that.

@Dorus then should we delete the license information here and make a header in every file?
not something we would do in this PR then

wonderingabout commented:
please do not merge yet, i need to rebase this in 1 commit with fixed commit notes

wonderingabout commented:
rebased with no changes since the last few days, i fixed what @Dorus mentionned here https://github.com/online-go/gtp2ogs/pull/260#issuecomment-626346293, are there other things to fix or we can merge this?

Dorus commented on connection.js:
```diff
@@ -173,6 +173,11 @@ class Connection {
             }
 
+            // Don't connect if it is not our turn.
+            if (gamedata.player_to_move !== this.bot_id) {
```
Why revert this change we made recently?

wonderingabout replied:
its a rebasing last minute mistake
thankfully you spotted it

fixed now :+1:

Dorus commented on connection.js:
```diff
@@ -305,7 +310,7 @@ class Connection {
             console.log("There are no connected games");
         }
-        const connected_games_per_user = this.countGamesForPlayer(notification.user.id);
+        const connected_games_per_user = this.gamesForPlayer(notification.user.id);
```
Why revert this change we made recently?

Dorus commented on connection.js:
```diff
@@ -500,10 +505,10 @@ class Connection {
-    //     const game = this.connectToGame(gamedata.id)
-    //     game.makeMove(gamedata.move_number);
-    // }
+    processMove(gamedata) {
```
unused function, why outcomment?

wonderingabout replied:
fixed, same as https://github.com/online-go/gtp2ogs/pull/260#discussion_r425345678

Dorus commented on connection.js:
```diff
@@ -535,7 +540,7 @@ class Connection {
         }
     }
-    countGamesForPlayer(player) {
+    gamesForPlayer(player) {
```
?

wonderingabout replied:
fixed, same as https://github.com/online-go/gtp2ogs/pull/260#discussion_r425345678

Dorus commented on pv.js:
```diff
@@ -12,6 +12,8 @@ class Pv {
             this.startupCheckSai = () => {}; // disable sai check for other bots.
         }
 
+        
+
```
Why add these newlines?

wonderingabout replied:
fixed, same as https://github.com/online-go/gtp2ogs/pull/260#discussion_r425345678

Dorus commented on pv.js:
```diff
@@ -37,6 +39,9 @@ class Pv {
                          }[setting];
         this.CLPV =      { 'PHOENIXGO':  (/\([^()]*\)/g) }[setting];
     }
+    fetchOgspvType() {
```
Is this function used? Why does this pr add code?

wonderingabout replied:
fixed, same as https://github.com/online-go/gtp2ogs/pull/260#discussion_r425345678

wonderingabout commented:
fixed all the last minute rebasing mistakes
thank you @Dorus for your vigilant eye, great! :100:  

should be ready to merge now :+1
and we can fix or enhance it in the future with this much better readme basis :+1:
